gflownet/envs/tetris.py

```diff
@@ -0,0 +1,489 @@
+"""
+An environment inspired by the game of Tetris.
+"""
+import itertools
+import re
+from typing import List, Optional, Tuple
+
+import numpy as np
+import numpy.typing as npt
+import torch
+from torchtyping import TensorType
+
+from gflownet.envs.base import GFlowNetEnv
+
+PIECES = {
+    "I": [1, [[1], [1], [1], [1]]],
+    "J": [2, [[0, 2], [0, 2], [2, 2]]],
+    "L": [3, [[3, 0], [3, 0], [3, 3]]],
+    "O": [4, [[4, 4], [4, 4]]],
+    "S": [5, [[0, 5, 5], [5, 5, 0]]],
+    "T": [6, [[6, 6, 6], [0, 6, 0]]],
+    "Z": [7, [[7, 7, 0], [0, 7, 7]]],
+}
+
+
+class Tetris(GFlowNetEnv):
+    """
+    Tetris environment: an environment inspired by the game of tetris. It's not
+    supposed to be a game, but rather a toy environment with an intuitive state and
+    action space.
+
+    The state space is 2D board, with all the combinations of pieces on it. Pieces are
```

Comment:
I didn't get what you mean here, especially by "an index that starts from piece_idx * max_pieces_per_type". It seems to me that indexes of pieces just go form 1 to 7, or maybe I'm missing something?

Reply:
Ah, I got what you mean after reading the tests and seeng numbers like 600 and 601 to identify different pieces of type 6. That makes sense (and it's also a nice solution for figuring out invalid backward actions easily). I'd paraphrase the docstring this way to make it more clear: "Pieces on the board are identified by a number that starts from piece_idx * max_pieces_per_type, and is incremented by 1 with each new piece from the same type. This number fills in the cells of the board where the piece is located"

Reply:
Yes, that's right. Sure, I'll do that, thanks!